File: akka-essentials/Chapter02/FirstAkkaApplicationScala/src/main/scala/akka/first/app/mapreduce/MapReduceApplication.scala
package akka.first.app.mapreduce

import akka.actor.{ActorSystem, Props}
import akka.first.app.mapreduce.actors.MasterActor
import akka.pattern.ask
import akka.util.Timeout
import scala.collection.immutable.Map
import scala.collection.mutable.ArrayBuffer
import scala.concurrent.Await
import scala.concurrent.duration._

sealed trait MapReduceMessage
case class WordCount(word: String, count: Int) extends MapReduceMessage
case class MapData(dataList: ArrayBuffer[WordCount]) extends MapReduceMessage
case class ReduceData(reduceDataMap: Map[String, Int]) extends MapReduceMessage
case class Result() extends MapReduceMessage

object MapReduceApplication {

  def main(args: Array[String]): Unit = {
    val _system = ActorSystem("MapReduceApp")
    val master = _system.actorOf(Props[MasterActor](), name = "master")
    master ! "The quick brown fox tried to jump over the lazy dog and fell on the dog"
    master ! "Dog is man's best friend"
    master ! "Dog and Fox belong to the same family"
      Thread.sleep(100)

    implicit val timeout = Timeout(Duration(5, "seconds"))
    val future = (master ? Result).mapTo[String]
    val result = Await.result(future, timeout.duration)
    println(result)

    terminate(_system)
  }

  private def terminate(system: ActorSystem): Unit =
    try {
      Thread.sleep(1000)
      println(">>> Press ENTER to exit <<<")
      System.in.read()
    }
    catch {
      case _: /*Interrupted,IO*/Exception => /* ignored */
    }
    finally {
      system.terminate()
    }

}

File: akka-essentials/Chapter02/FirstAkkaApplication/src/main/java/akka/first/app/mapreduce/MapReduceApplication.java
package akka.first.app.mapreduce;

import akka.actor.ActorRef;
import akka.actor.ActorSystem;
import akka.actor.Props;
import akka.first.app.mapreduce.actors.MasterActor;
import akka.first.app.mapreduce.messages.Result;
import akka.pattern.Patterns;
import akka.util.Timeout;
import java.util.concurrent.TimeUnit;
import scala.concurrent.Await;
import scala.concurrent.Future;
import scala.concurrent.duration.Duration;

public class MapReduceApplication {

    public static void main(String[] args) throws Exception {
        ActorSystem _system = ActorSystem.create("MapReduceApp");
        ActorRef master = _system.actorOf(
            Props.create(MasterActor.class),
            "master"
        );
        master.tell("The quick brown fox tried to jump over the lazy dog and fell on the dog", ActorRef.noSender());
        master.tell("Dog is man's best friend", ActorRef.noSender());
        master.tell("Dog and Fox belong to the same family", ActorRef.noSender());
        Thread.sleep(5000);

        Timeout timeout = new Timeout(Duration.create(5, TimeUnit.SECONDS));
        Future<Object> future = Patterns.ask(master, new Result(), timeout);
        String result = (String) Await.result(future, timeout.duration());
        System.out.println(result);

        _system.terminate();
    }

}
